core/src/services/fs/lister.rs

```diff
@@ -93,17 +103,61 @@ impl oio::BlockingList for FsLister<std::fs::ReadDir> {
         // (no extra system calls needed), but some Unix platforms may
         // require the equivalent call to symlink_metadata to learn about
         // the target file type.
-        let file_type = de.file_type().map_err(new_std_io_error)?;
 
-        let entry = if file_type.is_file() {
-            oio::Entry::new(&rel_path, Metadata::new(EntryMode::FILE))
-        } else if file_type.is_dir() {
+        let default_meta = self.op.metakey() == Metakey::Mode;
+
+        let (metadata, is_dir) = if default_meta {
+            let ft = de.file_type().map_err(new_std_io_error)?;
+            get_metadata(self.op.metakey(), None, Some(ft))?
+        } else {
+            let fs_meta = de.metadata().map_err(new_std_io_error)?;
+            get_metadata(self.op.metakey(), Some(fs_meta), None)?
+        };
+
+        let p = if is_dir {
             // Make sure we are returning the correct path.
-            oio::Entry::new(&format!("{rel_path}/"), Metadata::new(EntryMode::DIR))
+            &format!("{rel_path}/")
         } else {
-            oio::Entry::new(&rel_path, Metadata::new(EntryMode::Unknown))
+            &rel_path
         };
 
-        Ok(Some(entry))
+        Ok(Some(oio::Entry::new(p, metadata)))
     }
 }
+
+fn get_metadata(
+    metakey: FlagSet<Metakey>,
+    fs_meta: Option<std::fs::Metadata>,
+    file_type: Option<FileType>,
+) -> Result<(Metadata, bool)> {
+    let ft = if let Some(t) = fs_meta.as_ref() {
+        t.file_type()
+    } else if let Some(t) = file_type {
+        t
+    } else {
+        panic!("At least one metadata should be provided!")
+    };
+
+    let mut meta = if ft.is_file() {
+        Metadata::new(EntryMode::FILE)
+    } else if ft.is_dir() {
+        Metadata::new(EntryMode::DIR)
+    } else {
+        Metadata::new(EntryMode::Unknown)
+    };
+
+    if fs_meta.is_some() {
+        let contains_metakey = |k| metakey.contains(k) || metakey.contains(Metakey::Complete);
+        let fs_meta = fs_meta.unwrap();
+
+        if contains_metakey(Metakey::ContentLength) {
```

Comment:
We don't need to check the meta key. If we have sent `metadata` syscall, we can fill in any known data.